Mathlib/Analysis/Convex/Quasiconvex.lean

```diff
@@ -87,6 +88,105 @@ theorem QuasiconcaveOn.convex [IsDirected β (· ≥ ·)] (hf : QuasiconcaveOn 
 
 end LE_β
 
+section Composition
+
+variable {𝕜 E β : Type*} [Semiring 𝕜] [PartialOrder 𝕜]
+  [AddCommMonoid E] [SMul 𝕜 E]
+
+variable {β γ : Type*} [LinearOrder β] [Preorder γ]
+
+variable {s : Set E} {f : E → β} {g : β → γ}
+
+theorem QuasiconvexOn.monotone_comp
+    (hg : Monotone g) (hf : QuasiconvexOn 𝕜 s f) :
+    QuasiconvexOn 𝕜 s (g ∘ f) := fun c x hx y hy ↦ by
+  simp only [Function.comp_apply, mem_setOf_eq] at hx hy
+  intro a b ha hb hab
+  simp only [Function.comp_apply, mem_setOf_eq]
+  cases le_total (f x) (f y) with
+  | inl h =>
+      specialize hf (f y) ⟨hx.1, h⟩ ⟨hy.1, le_rfl⟩ ha hb hab
+      simp only [mem_setOf_eq] at hf
+      exact ⟨hf.1, le_trans (hg hf.2) hy.2⟩
+  | inr h =>
+      specialize hf (f x) ⟨hx.1, le_rfl⟩ ⟨hy.1, h⟩ ha hb hab
+      simp only [mem_setOf_eq] at hf
+      exact ⟨hf.1, le_trans (hg hf.2) hx.2⟩
+
+theorem QuasiconvexOn.antitone_comp (hg : Antitone g) (hf : QuasiconvexOn 𝕜 s f) :
+    QuasiconcaveOn 𝕜 s (g ∘ f) :=
+  hf.monotone_comp (γ := γᵒᵈ) hg
+
+theorem QuasiconcaveOn.monotone_comp (hg : Monotone g) (hf : QuasiconcaveOn 𝕜 s f) :
+    QuasiconcaveOn 𝕜 s (g ∘ f) :=
+  QuasiconvexOn.monotone_comp hg.dual hf
+
+theorem QuasiconcaveOn.antitone_comp (hg : Antitone g) (hf : QuasiconcaveOn 𝕜 s f) :
+    QuasiconvexOn 𝕜 s (g ∘ f) :=
+  QuasiconvexOn.antitone_comp hg.dual hf
+
+theorem QuasilinearOn.monotone_comp (hg : Monotone g) (hf : QuasilinearOn 𝕜 s f) :
+    QuasilinearOn 𝕜 s (g ∘ f) :=
+  ⟨hf.1.monotone_comp hg, hf.2.monotone_comp hg⟩
+
+theorem QuasilinearOn.antitone_comp (hg : Antitone g) (hf : QuasilinearOn 𝕜 s f) :
+    QuasilinearOn 𝕜 s (g ∘ f) :=
+  ⟨hf.2.antitone_comp hg, hf.1.antitone_comp hg⟩
+
+end Composition
+
+section Restriction
+
+variable {𝕜 E : Type*} [Semiring 𝕜] [PartialOrder 𝕜]
+  [AddCommMonoid E] [SMul 𝕜 E]
+variable {β : Type*} [Preorder β]
+variable {s : Set E} {f : E → β}
+
+theorem QuasiconvexOn.mono {t : Set E}
+    (hf : QuasiconvexOn 𝕜 s f) (hst : t ⊆ s) (ht : Convex 𝕜 t) :
+    QuasiconvexOn 𝕜 t f := by
+  intro b
+  rw [Set.sep_of_subset hst]
+  exact Convex.inter ht (hf b)
+
+theorem QuasiconcaveOn.mono {t : Set E}
+    (hf : QuasiconcaveOn 𝕜 s f) (hst : t ⊆ s) (ht : Convex 𝕜 t) :
+    QuasiconcaveOn 𝕜 t f := by
+  intro b
+  rw [Set.sep_of_subset hst]
+  exact Convex.inter ht (hf b)
+
+end Restriction
+
+section Preconnected
+
+variable {E : Type*} [AddCommGroup E] [Module ℝ E]
+  [TopologicalSpace E] [IsTopologicalAddGroup E] [ContinuousSMul ℝ E]
+
+variable {β : Type*} [Preorder β] {f : E → β}
+
+open scoped Set.Notation
+
+theorem QuasiconcaveOn.isPreconnected_preimage {s : Set E} {t : β}
```

Comment:
The lemma name does not really indicate that you're looking inside the subtype (given the name, I'd expect more something saying `IsPreconnected (s ∩ f ⁻¹' Ici t)`, i.e., the definition of quasiconcave). Maybe rename it to
```suggestion
theorem QuasiconcaveOn.isPreconnected_preimage_subtype {s : Set E} {t : β}
```
Same thing below.